### PR TITLE
Implement QueryOp class.

### PR DESCRIPTION
**Pull request by mqliang**
## Description
Implement QueryOp class, which take a file that has queries, and a file that has results, run the queries through the broker/server and compare the results with the result file.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

**Comment by mqliang** [comment]
@mcvsubbu

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java [hunk 1]
```diff
@@ -36,8 +36,7 @@ private CompatibilityOpsRunner(String configFileName) {
     _configFileName = configFileName;
   }
 
-  private boolean runOps()
-      throws IOException, JsonParseException, JsonMappingException {
+  private boolean runOps() throws IOException, JsonParseException, JsonMappingException, InterruptedException {
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:39]: (nit) Please set up the Pinot Code Style following the instructions [here](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup) and reformat the changes

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java [hunk 2]
```diff
@@ -52,6 +51,11 @@ private boolean runOps()
         System.out.println("Failure");
         break;
       }
+      /*
+       Brokers take time to update route table after a segment is created/uploaded/deleted, so sleep 1s between
+       two operations
+       */
+      TimeUnit.SECONDS.sleep(1);
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:58]: Adding a sleep here can make tests unreliable and susceptible to intermittent failures. You should add logic like the following:
`
ready = test_broker();
while (!ready) {
  sleep(200ms)
  ready = test_broker();
}
`
In test_broker(), you can do one of:
(1) make a call to the debug port of the broker to ensure that routing tables are built
(2) send a query to verify

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java (hunk 1 above)
Comment (mcvsubbu) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:39]: +1

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 3]
```diff
@@ -30,13 +49,34 @@
...
   public QueryOp() {
     super(OpType.QUERY_OP);
   }
 
+  private static boolean shouldIgnore(String line) {
+    String trimmedLine = line.trim();
+    return trimmedLine.isEmpty() || trimmedLine.startsWith("#");
+  }
+
+  private static String constructResponse(InputStream inputStream) throws IOException {
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:69]: why static?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 4]
```diff
@@ -30,13 +49,34 @@
...
+  private static final String TIME_USED_MS = "timeUsedMs";
+  private final String _brokerBaseApiUrl = ClusterDescriptor.CONTROLLER_URL;
   private String _queryFileName;
   private String _expectedResultsFileName;
 
   public QueryOp() {
     super(OpType.QUERY_OP);
   }
 
+  private static boolean shouldIgnore(String line) {
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:64]: why static?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java (hunk 2 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:58]: Both (1) and (2) need hardcode table name in code, which is not good in my opinion. Another idea is: in QueryOp.runOps() function, when test the first query, if the error message is "{errorCode: 410, message: "BrokerResourceMissingError"}", sleep and retry, until it succeeded or timeout.

Comment (mcvsubbu) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:58]: Wait. I am not sure why the sleep or any kind of wait is needed here at all. Let us discuss.

**Comment by mqliang** [comment]
I have address the comments:
1. reformat the code
2. add a "verifyRoutingTableUpdated" method in SegmentOp class instead of sleep between two Op
3. move send query related code to a dedicated QueryProcessor class, since SegmentOp also need to send query to verify routing table is updated.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java [hunk 5]
```diff
@@ -65,12 +66,6 @@
   private static final FileFormat DEFAULT_FILE_FORMAT = FileFormat.CSV;
   private static final int DEFAULT_MAX_SLEEP_TIME_MS = 30000;
   private static final int DEFAULT_SLEEP_INTERVAL_MS = 200;
-
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java:68]: Why did you move this?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 6]
```diff
@@ -30,13 +38,22 @@
  */
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class QueryOp extends BaseOp {
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryOp.class);
+
+  private static final String NUM_DOCS_SCANNED = "numDocsScanned";
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:43]: Rename it to `NUM_DOCS_SCANNED_KEY`. Same for the following one.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 7]
```diff
@@ -30,13 +38,22 @@
...
   private String _queryFileName;
   private String _expectedResultsFileName;
 
   public QueryOp() {
     super(OpType.QUERY_OP);
   }
 
+  private boolean shouldIgnore(String line) {
+    String trimmedLine = line.trim();
+    return trimmedLine.isEmpty() || trimmedLine.startsWith("#");
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:54]: Put "#" into a constant.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 8]
```diff
@@ -55,7 +72,97 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
 
   @Override
   boolean runOp() {
-    System.out.println("Verifying queries in " + _queryFileName + " against results in " + _expectedResultsFileName);
-    return true;
+    System.out.printf("Verifying queries in %s against results in %s\n", _queryFileName, _expectedResultsFileName);
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:75]: Why not use the `LOGGER` you specified in Line 41?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 9]
```diff
@@ -55,7 +72,97 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+  }
+
+  boolean verifyQueries()
+      throws Exception {
+    BufferedReader expectedResultReader = null;
+    boolean testPassed = false;
+
+    try (BufferedReader queryReader = new BufferedReader(
+        new InputStreamReader(new FileInputStream(_queryFileName), StandardCharsets.UTF_8))) {
+      if (_queryFileName == null) {
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:91]: I'm not sure whether this if statement can ever be satisfied. Maybe we should check that before the try block?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 10]
```diff
@@ -55,7 +72,97 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+
+        JsonNode expectedJson = null;
+        try {
+          String expectedResultLine = expectedResultReader.readLine();
+          while (shouldIgnore(expectedResultLine)) {
+            expectedResultLine = expectedResultReader.readLine();
+          }
+          expectedJson = JsonUtils.stringToJsonNode(expectedResultLine);
+        } catch (Exception e) {
+          LOGGER.error("Comparison FAILED: Line: {} Exception caught while getting expected response for query: '{}'",
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:118]: If an exception was caught when parsing the expected results, should we skip firing queries to the cluster and stop the comparison immediately?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 11]
```diff
@@ -55,7 +72,97 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+    BufferedReader expectedResultReader = null;
+    boolean testPassed = false;
+
+    try (BufferedReader queryReader = new BufferedReader(
+        new InputStreamReader(new FileInputStream(_queryFileName), StandardCharsets.UTF_8))) {
+      if (_queryFileName == null) {
+        LOGGER.error("Result file is missing!");
+        return testPassed;
+      } else {
+        expectedResultReader = new BufferedReader(
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:95]: I think you can put this to the same try bracket in  Line 89 as well, so that you don't need to specify the finally block at the end.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryProcessor.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Map;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+public class QueryProcessor {
+
+  private static final String _brokerBaseApiUrl = ClusterDescriptor.CONTROLLER_URL;
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryProcessor.java:39]: Use snake case instead of camel case for static constant?

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java [hunk 13]
```diff
@@ -241,6 +242,27 @@ private boolean verifySegmentInState(String state)
     return true;
   }
 
+  private boolean verifyRoutingTableUpdated()
+      throws Exception {
+    String query = "SELECT count(*) FROM " + _tableName;
+    JsonNode result = QueryProcessor.postSqlQuery(query);
+    System.out.println(result);
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java:249]: remove this System.out.println()?

**Comment by mcvsubbu** [comment]
Issue #4854

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 8 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:75]: Just  wanner make it consistence with StreamOp and TableOp, basically every Op will print out such a info message in console.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 10 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:118]: From all-queries point of view, there is no early-termination logic here. Say we have 100 queries, if the first query failed, instead of returning test failure immediately, we will continue to verify the remaining 99 queries, and we will return test failure at the very end since 99 != 100.  In this way, can detect as many problematic queries as possible in one go.

From a single query point of view, yes if an exception was caught when parsing the expected results, we indeed skip firing queries to the cluster:
```
if (expectedJson != null) {
          try {
            actualJson = QueryProcessor.postSqlQuery(query);
          } catch (Exception e) {
            LOGGER.error("Comparison FAILED: Line: {} Exception caught while running query: '{}'", queryLineNum, query,
                e);
          }
        }
```
We only fire query to cluster if `expectedJson != null`

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java (hunk 13 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 8 above)
Comment (mcvsubbu) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:76]: For now, it is ok.  For some reason , logs don't show up when we run the ops via command-line. Eventually we will remove the sout or keep a small number of them if needed.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 11 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 9 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: Use a try-with-resources statement now.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 6 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 14]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+  boolean verifyQueries()
+      throws Exception {
+    boolean testPassed = false;
+
+    try (BufferedReader queryReader = new BufferedReader(
+        new InputStreamReader(new FileInputStream(_queryFileName), StandardCharsets.UTF_8));
+        BufferedReader expectedResultReader = new BufferedReader(
+            new InputStreamReader(new FileInputStream(_expectedResultsFileName), StandardCharsets.UTF_8))) {
+
+      int passed = 0;
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:94]:
```suggestion
      int SucceededQueryCount = 0;
```

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 15]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+      int total = 0;
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:95]:
```suggestion
      int totalQueryCount = 0;
```

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 16]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+            actualJson = QueryProcessor.postSqlQuery(query);
+          } catch (Exception e) {
+            LOGGER.error("Comparison FAILED: Line: {} Exception caught while running query: '{}'", queryLineNum, query,
+                e);
+          }
+        }
+
+        if (expectedJson != null && actualJson != null) {
+          try {
+            boolean comparisonResult = SqlResultComparator.areEqual(actualJson, expectedJson, query);
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:129]:
```suggestion
            boolean passed = SqlResultComparator.areEqual(actualJson, expectedJson, query);
```

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 17]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+            if (comparisonResult) {
+              passed++;
+              LOGGER.info("Comparison PASSED: Line: {} actual Time: {} ms expected Time: {} ms Docs Scanned: {}",
+                  queryLineNum, actualJson.get(TIME_USED_MS_KEY), expectedJson.get(TIME_USED_MS_KEY),
+                  actualJson.get(NUM_DOCS_SCANNED_KEY));
+              LOGGER.debug("actual Response: {}", actualJson);
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:135]: Please combine these two logs into one log.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 18]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+                e);
+          }
+        }
+
+        if (expectedJson != null && actualJson != null) {
+          try {
+            boolean comparisonResult = SqlResultComparator.areEqual(actualJson, expectedJson, query);
+            if (comparisonResult) {
+              passed++;
+              LOGGER.info("Comparison PASSED: Line: {} actual Time: {} ms expected Time: {} ms Docs Scanned: {}",
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:132]: We should not be logging expected time, etc. We should not even have it in metadata. We can add some outliers for time later. For now, remove it from the results file and also from this log.
Also, we don't need an INFO log for every passing query. Just add a single  debug log with query, expected result and actual result

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java [hunk 19]
```diff
@@ -55,7 +73,87 @@ public void setExpectedResultsFileName(String expectedResultsFileName) {
...
+                  queryLineNum, actualJson.get(TIME_USED_MS_KEY), expectedJson.get(TIME_USED_MS_KEY),
+                  actualJson.get(NUM_DOCS_SCANNED_KEY));
+              LOGGER.debug("actual Response: {}", actualJson);
+              LOGGER.debug("expected Response: {}", expectedJson);
+            } else {
+              LOGGER.error("Comparison FAILED: Line: {} query: {}", queryLineNum, query);
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:138]: Single log.error will suffice.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java [hunk 20]
```diff
@@ -67,8 +68,7 @@
   private static final int DEFAULT_SLEEP_INTERVAL_MS = 200;
 
   public enum Op {
-    UPLOAD,
-    DELETE
+    UPLOAD, DELETE
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java:71]: Keep them in separate lines please

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java [hunk 21]
```diff
@@ -241,6 +242,26 @@ private boolean verifySegmentInState(String state)
     return true;
   }
 
+  private boolean verifyRoutingTableUpdated()
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java:245]: You should really be getting the number of rows before adding the segment, and then the number of rows after adding the segment and make sure that it has increased by the number of rows in the segment. If you do this, then you can remove the other check of segment state model.online.

You can mark this as TODO and take it on in a separate PR

**Review by mcvsubbu — COMMENTED** [review]
Mostly minor comments, otherwise lgtm

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java (hunk 21 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java:248]: marked as TODO.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 14 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 16 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 18 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java (hunk 20 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java [hunk 22]
```diff
@@ -36,8 +35,7 @@ private CompatibilityOpsRunner(String configFileName) {
     _configFileName = configFileName;
   }
 
-  private boolean runOps()
-      throws IOException, JsonParseException, JsonMappingException {
+  private boolean runOps() throws IOException, JsonParseException, JsonMappingException, InterruptedException {
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java:38]: I'd suggest throwing a generic `Exception` instead of putting all the concrete ones here.

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 8 above)
Comment (jackjlli) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:76]: Then why use `printf` instead of `println` here?

**Review by jackjlli — APPROVED** [review]
Minor but LGTM. Thanks for working on this!

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java (hunk 22 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/CompatibilityOpsRunner.java]: done

File: pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java (hunk 8 above)
Comment (mcvsubbu) [reply pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/QueryOp.java:76]: As suggested, please change this to println instead of printf. thanks

**Review by mcvsubbu — APPROVED** [review]
